Code/src/main/java/nl/utwente/group10/haskell/type/FuncT.java

```diff
@@ -30,4 +30,10 @@ public final String toHaskellType() {
         out.append(")");
         return out.toString();
     }
+
+    @Override
+    public FuncT getFresh() {
+        Type[] args = getFreshArgs();
+        return new FuncT(args[0], args[1]);
+    }
```

Comment:
Hmm, no, I don't think you would do instanceof in `ConstT.toString`, that would be the wrong way around: the idea is to make the object identity less important, not more important. You'd switch on `this.constructor` instead, I think.

The current set-up would be very Java-like if ConstT was abstract, but it isn't, so you have FuncT's which are somehow more special but still compare equal to plain ConstT's:

````
    @Test
    public final void testNaked() {
        ConstT u = new ConstT("Unit");

        // A function type is a constant type with a constructor of ->
        ConstT f = new ConstT("->", u, u);

        // You can also write that as new FuncT(u, u)
        FuncT ft = new FuncT(u, u);

        // As you can see, these two types are completely equal (compareTo == 0)
        assertEquals(0, f.compareTo(ft));

        // Except that they aren't
        assertEquals(f.toHaskellType(), ft.toHaskellType());
    }
````

The factory method sounds like a neat idea. You could have it do some FuncT-specific trickery as well.